Add Eigen linear algebra library

calcmogul opened the pull request:
This imports Eigen 3.3.7 with a patch applied to make it build with GCC
9. This will be used by the wpilibc implementation of state-space
control and mecanum/swerve forward kinematics (the forward kinematics
requires least-squares solutions via a matrix pseudoinverse).

Eigen is licensed under BSD, MINPACK, and MPL2. The license included is
BSD.

PeterJohnson commented:
Why the separation of Eigen and unsupported/Eigen?

calcmogul commented:
That's how Eigen organizes them. A typical Eigen install has the following structure:
```
[tav@myriad ~]$ find /usr/include/eigen3 -maxdepth 2 -type d
/usr/include/eigen3
/usr/include/eigen3/unsupported
/usr/include/eigen3/unsupported/Eigen
/usr/include/eigen3/Eigen
/usr/include/eigen3/Eigen/src
```
We could stick everything in an `eigen3` directory if we can add a `-I` directive somewhere. pkg-config lets you use `Eigen/Core` instead of `eigen3/Eigen/Core`.

calcmogul commented:
I'm not sure why Eigen has an unsupported directory in the first place though. My guess is it's their version of a contrib directory. The state-space PR uses the matrix exponential from the unsupported directory in the PeriodVariantPlant class.

PeterJohnson commented:
Ok. In our case (at least for cmake installation), everything in the include directory gets installed to a wpiutil subdirectory, so we don't want to add another level.

PeterJohnson commented:
I would prefer to do this as two commits (e.g. upstream and gcc9 patches) if that's not too difficult to do.

ThadHouse commented:
And special casing Gradle to add a new include path is not easy either. So everything rooted to the main include root is pretty much the only way for us.

calcmogul commented:
I have a copy of the patch, so that should be easy.

calcmogul commented:
The patch for GCC 9 is now in the second commit.

PeterJohnson commented:
https://github.com/eigenteam/eigen-git-mirror/blob/master/COPYING.README says that most of the code is MPL2, and you have to define a preprocessor symbol to limit it to just MPL2 (and not include LGPL code).

calcmogul commented:
I'll work on doing that tomorrow night.

calcmogul commented:
Are we allowed to have LGPL code in the codebase if it doesn't get compiled in?

ThadHouse commented:
It doesn't actually look like any LGPL code exists in that repo. The symbol `EIGEN_MPL2_ONLY` is only used from 2 files, and both are BSD. 

There is a GPL file `unsupported/test/mpreal/mpreal.h`, but it looks like its easy to just remove that.

calcmogul commented:
The version of Eigen imported here doesn't have a test folder or mpreal.h in it. I copied it out of my /usr/include folder and it was installed with the following:
```
build() {
	mkdir -p build
	cd build
	cmake ../eigen-eigen-* \
		-DCMAKE_BUILD_TYPE=Release \
		-DCMAKE_INSTALL_PREFIX=/usr \
                -DCMAKE_INSTALL_LIBDIR=lib
}

package() {
	cd build
	make DESTDIR="$pkgdir" install

	# install custom licenses
	install -Dm644 ../eigen-eigen-*/COPYING.MPL2 "$pkgdir/usr/share/licenses/$pkgname/LICENSE"
        install -Dm644 ../eigen-eigen-*/COPYING.BSD "$pkgdir/usr/share/licenses/$pkgname/LICENSE"
        install -Dm644 ../eigen-eigen-*/COPYING.MINPACK "$pkgdir/usr/share/licenses/$pkgname/LICENSE"
}
```

calcmogul commented:
I removed vectorization support for platforms we don't use and changed the license included to MPL2, since the only source files left have that license.
